apiserver/facades/client/action/operation.go

```diff
@@ -5,94 +5,170 @@ package action
-	if err := a.checkCanRead(ctx); err != nil {
-		return params.OperationResults{}, errors.Trace(err)
 	}
 
-	var receiverTags []names.Tag
-	for _, name := range arg.Units {
-		receiverTags = append(receiverTags, names.NewUnitTag(name))
+	// If no valid run params (all receivers invalid), do not call service; return per-action errors.
+	if len(runParams) == 0 {
+		return params.EnqueuedActions{Actions: actionResults}, nil
 	}
-	for _, id := range arg.Machines {
-		receiverTags = append(receiverTags, names.NewMachineTag(id))
+
+	result, err := a.operationService.Run(ctx, runParams)
+	if err != nil {
+		return params.EnqueuedActions{}, errors.Capture(err)
 	}
 
-	var unitNames []coreunit.Name
-	if len(arg.ActionNames) == 0 && len(arg.Applications) == 0 && len(receiverTags) == 0 {
-		var err error
-		unitNames, err = a.applicationService.GetAllUnitNames(ctx)
-		if err != nil {
-			return params.OperationResults{}, errors.Trace(err)
+	for _, unitResult := range result.Units {
+		targetName := unitResult.ReceiverName.String()
+		if unitResult.IsLeader {
+			targetName = unitResult.ReceiverName.Application() + leader
 		}
-	} else {
-		for _, aName := range arg.Applications {
-			appUnitName, err := a.applicationService.GetUnitNamesForApplication(ctx, aName)
-			if errors.Is(err, applicationerrors.ApplicationNotFound) {
-				return params.OperationResults{}, errors.NotFoundf("application %q", aName)
-			} else if err != nil {
-				return params.OperationResults{}, errors.Trace(err)
-			}
-			unitNames = append(unitNames, appUnitName...)
+		ar := actionResultByUnitName[targetName]
+		if ar == nil {
+			return params.EnqueuedActions{}, errors.Errorf("unexpected result for %q", targetName)
 		}
+		*ar = toActionResult(names.NewUnitTag(unitResult.ReceiverName.String()), unitResult.TaskInfo)
 	}
-	for _, unitName := range unitNames {
-		tag := names.NewUnitTag(unitName.String())
-		receiverTags = append(receiverTags, tag)
+	// Mark missing results
+	for key, ar := range actionResultByUnitName {
+		if ar != nil && ar.Action == nil && ar.Error == nil {
+			ar.Error = apiservererrors.ServerError(jujuerrors.NotFoundf("missing result for %q", key))
+		}
+	}
+	return params.EnqueuedActions{
+		OperationTag: names.NewOperationTag(result.OperationID).String(),
+		Actions:      actionResults,
+	}, nil
+}
+
+// ListOperations fetches the called actions for specified apps/units.
+func (a *ActionAPI) ListOperations(ctx context.Context, arg params.OperationQueryArgs) (params.OperationResults, error) {
+	if err := a.checkCanRead(ctx); err != nil {
+		return params.OperationResults{}, errors.Capture(err)
+	}
+
+	args := operation.QueryParams{
+		Target:      makeOperationTarget(arg.Applications, arg.Machines, arg.Units),
+		ActionNames: arg.ActionNames,
+		Status:      arg.Status,
+		Limit:       arg.Limit,
+		Offset:      arg.Offset,
+	}
+	result, err := a.operationService.GetOperations(ctx, args)
```

Comment:
This makes sense a a bulk call into the domain. Neither `Run` nor `Operation` do looking at the code.

Reply:
Looking at the 3.6 code, it wasn't done well there. Perhaps it's okay here. Client facades will be rewritten. The trick will be ensuring that the order isn't changed.